### PR TITLE
#179 alarm manager

### LINKED ISSUE #179: DS-179: As a developer, I would like to create a functionality that will automatically start the checking of the supervised user attendance status around the time of the scheduled event
This issue is only supposed to start the function / feature implemented in issue # 177 , not to implement how to get the user's location as that will be done in issue #177. This issue will request the user attendance status from functionality in issue # 177 as from about 5 min before the beginning of the scheduled event and at every consecutive interval of time until the user's attendance status is given as present or until the event time is passed by 1 hour. The consecutive interval of time can be every 5 min. In case an attendance status is obtained as "not detected ", a text sms should be sent to the supervised user to let him/ her know about that. In case the event time is passed by 1 hour, a text sms should be sent to the supervised user to let her / him that she / he had  been reported as absent and that he / she should let the monitoring user / supervisor know if she/ he arrives at a later time

### PR DESCRIPTION
**Pull request by ShawnGuerra-Bautista**
### Related Issues
#179

### Description
Creates an alarms for each schedule in which the activation time is set to the start time of the schedule. The alarms work even if the app is closed or the phone is sleeping. 

The alarms are set each time you visit the homepage. In practice, this is important since, if there's a new schedule in the database, a new alarm will be set for that schedule. Also, because of `PendingIntent.FLAG_UPDATE_CURRENT`, the old alarms will kept, but will update the PendingIntent of the alarms if there are any changes to it.

### How to check PR
This is all back end stuff, so there's no way to test this feature with UI functionality without "bloating" it with code that will be removed later. So, run the app, visit the home page once, and then you will receive a notification after 30 seconds even when the app is closed, and/or the phone is sleeping. 

The notification should have as description: Hello there 10 -82. Which represents the uid, the latitude, and longitude of the schedule.

NOTE: If you've visited the home page multiple time, then you may receive multiple notifications. That's because the "testing" code create an alarm a 30 seconds after the current time (in milliseconds).

Please check the following files: AlarmBroadcastReceiver.java, and AlarmBroadcastReceiverManager.java.
Only check code without the comments
`// TODO: --START-- REMOVE BEFORE MERGING PULL REQUEST. THIS IS ONLY FOR TESTING PURPOSES.`
`// TODO: --END-- REMOVE BEFORE MERGING PULL REQUEST. THIS IS ONLY FOR TESTING PURPOSES.`
 around it because these pieces of code are only to "visualize" the AlarmManager working. They will be removed later.

NOTE: The code smells from SonarQube is due to the testing code, so do not worry about it; they will be eliminated after I remove them.

### CODE REVIEW

**Review by KennanHoa — APPROVED** [review]
Alarm works as intended (received notification after 30 seconds)
Clean codes without any unnecessary lines
Clear comments and documentations

File: app/src/main/java/com/example/bigowlapp/activity/HomePageActivity.java
Comment (refatK) [inline app/src/main/java/com/example/bigowlapp/activity/HomePageActivity.java]: Good use of a manager so that the activity isn't bloated, and we can easily move it to other places if needed

File: app/src/main/java/com/example/bigowlapp/repository/ScheduleRepository.java
Comment (refatK) [inline app/src/main/java/com/example/bigowlapp/repository/ScheduleRepository.java:19]: Constants are always good, so nice refactor

File: app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java
Comment (refatK) [inline app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java]: What type of error would show here btw. If it would be confusing to the user, might be better to simply Log the error, but depends what a failure results in.

File: app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java
Comment (refatK) [inline app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java]: This can be refactored using [Replace Loop with Pipeline](https://refactoring.com/catalog/replaceLoopWithPipeline.html) using java 8 stream.

Although I think this is ok too tbh.

File: app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java
Comment (refatK) [inline app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java]: can use OnSuccessTask to reduce complexity (as the if statement would be unnecessary), but not necessary change.

File: app/src/main/java/com/example/bigowlapp/utils/IntentConstants.java
Comment (refatK) [inline app/src/main/java/com/example/bigowlapp/utils/IntentConstants.java:3]: Good way of storing intent constants

File: app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java
Comment (refatK) [inline app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java]: Might want to consider a name change to reflect the specific puprose of this alarm manager, especially in the case we need alarm for other things. Name it based on the purpose of this specific Alarm manager.

File: app/src/main/java/com/example/bigowlapp/service/AlarmBroadcastReceiver.java
Comment (refatK) [inline app/src/main/java/com/example/bigowlapp/service/AlarmBroadcastReceiver.java]: Might want to consider a name change to reflect the specific purpose of this alarm BroadcastReciever, especially in the case we need alarm receivers for other things. Name it based on the purpose of this specific alarm BroadcastReciever.

**Review by refatK — CHANGES_REQUESTED** [review]
Pretty good, will also test the functionality itself.

I've asked for some changes and also some suggestions, but it's all pretty minor.

File: app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java
Comment (ShawnGuerra-Bautista) [reply app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java]: Yes, it would reduce the complexity.

Problem is that onSuccessTask doesn't return an exception, so if there's a problem/error, there won't be a way to handle it.

I'm not sure if this is a function that doesn't require exception handling.

Comment (refatK) [reply app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java]: @ShawnGuerra-Bautista It actually does, you just don't need to specifically define it.

As for the disadvantage of OnSuccessTask, if you need to do something other than throw an error on failure, you can't. Although looking at the code right now, the else statement only throws the error, hence my suggestion of OnSuccessTask.

Behavior should be exactly the same.

**Comment by refatK** [comment]
Just tested the functionalty and works as expected. Duplicate notifications also don't get created if they already exist as you suggested in your description with use of `PendingIntent.FLAG_UPDATE_CURRENT`.

Good job!

File: app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java
Comment (ShawnGuerra-Bautista) [reply app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java]: Although the performance will be worst according to this source (https://stackoverflow.com/questions/44180101/in-java-what-are-the-advantages-of-streams-over-loops), it's nice for readability and I don't think performance will take a significant hit, so I'll do the refactor.

File: app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java
Comment (ShawnGuerra-Bautista) [reply app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java]: https://developers.google.com/android/reference/com/google/android/gms/tasks/Task#onSuccessTask(com.google.android.gms.tasks.SuccessContinuation

Alright, I re-read it, and I believe you are right. So, I'll do the changes.

File: app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java
Comment (refatK) [reply app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java]: I actually didn't know that, good to know.

File: app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java
Comment (ShawnGuerra-Bautista) [reply app/src/main/java/com/example/bigowlapp/utils/AlarmBroadcastReceiverManager.java]: I would display an error concerning querying schedules.

In practice, this error shouldn't be reached, but, for some exceptional cases that Firestore throws an error, the error that is thrown by Firestore will be displayed to the user. I believe that Firestore would throw errors with useful message.

**Comment by refatK** [comment]
Thanks for addressing my request. Once you remove the test code, I'll approve

**Review by refatK — APPROVED** [review]
Everything has been addressed, so approving now!

**Pull request merged**
